src/coreclr/jit/valuenum.cpp

```diff
@@ -8492,7 +8492,98 @@ void Compiler::fgValueNumberSsaVarDef(GenTreeLclVarCommon* lcl)
     }
 }
 
-void Compiler::fgValueNumberTree(GenTree* tree)
+//------------------------------------------------------------------------
+// fgValueNumberConstStringElemLoad: Try to match "cns_str"[cns_index] tree
+//    and apply a constant VN representing given char at cns_index in that string.
+//
+// Arguments:
+//    tree - the GT_IND node
+//
+// Return Value:
+//    true if the pattern was recognized and a new VN is assigned
+//
+bool Compiler::fgValueNumberConstStringElemLoad(GenTreeIndir* tree)
+{
+    ValueNum  addrVN = tree->gtGetOp1()->gtVNPair.GetLiberal();
+    VNFuncApp funcApp;
+    if (!varTypeIsShort(tree) || !vnStore->GetVNFunc(addrVN, &funcApp))
+    {
+        return false;
+    }
+
+    // Is given VN representing a frozen object handle
+    auto isCnsObjHandle = [](ValueNumStore* vnStore, ValueNum vn, CORINFO_OBJECT_HANDLE* handle) -> bool {
+        if (vnStore->IsVNHandle(vn) && (vnStore->GetHandleFlags(vn) == GTF_ICON_OBJ_HDL))
+        {
+            const size_t obj = vnStore->CoercedConstantValue<size_t>(vn);
+            *handle          = reinterpret_cast<CORINFO_OBJECT_HANDLE>(obj);
+            return true;
+        }
+        return false;
+    };
+
+    CORINFO_OBJECT_HANDLE objHandle = nullptr;
+    int                   index     = -1;
+
+    // First, let see if we have PtrToArrElem
+    ValueNum addr = funcApp.m_args[0];
+    if (funcApp.m_func == VNF_PtrToArrElem)
+    {
+        ValueNum arrVN  = funcApp.m_args[1];
+        ValueNum inxVN  = funcApp.m_args[2];
+        ssize_t  offset = vnStore->ConstantValue<ssize_t>(funcApp.m_args[3]);
+
+        if (isCnsObjHandle(vnStore, arrVN, &objHandle) && offset == 0 && vnStore->IsVNConstant(inxVN))
+        {
+            index = (int)vnStore->CoercedConstantValue<size_t>(inxVN);
+        }
+    }
+    else if (((genTreeOps)funcApp.m_func == GT_ADD) && vnStore->IsVNConstant(funcApp.m_args[0]))
+    {
+        // Here we try to match these two patterns:
+        //
+        //  1) (objHandle + firstCharOffset)
+        //  2) ((objHandle + firstCharOffset) + byteIndex)
+        //
+        size_t dataOffset = 0;
+        if (vnStore->IsVNConstant(funcApp.m_args[1]) && isCnsObjHandle(vnStore, funcApp.m_args[0], &objHandle))
+        {
+            dataOffset = vnStore->CoercedConstantValue<size_t>(funcApp.m_args[1]);
+        }
+        else
+        {
+            assert(vnStore->IsVNConstant(funcApp.m_args[0]));
+            const size_t byteDataIndex = vnStore->CoercedConstantValue<size_t>(funcApp.m_args[0]);
+            if (vnStore->GetVNFunc(funcApp.m_args[1], &funcApp) && ((genTreeOps)funcApp.m_func == GT_ADD) &&
+                vnStore->IsVNConstant(funcApp.m_args[1]) && isCnsObjHandle(vnStore, funcApp.m_args[0], &objHandle))
```

Comment:
VN canonicalization means we usually have to check for both `m_args[0]` and `m_args[1]` to be constant.

Reply:
um.. isn't m_args[0] is checked via `isCnsObjHandle` here?

Reply:
What I mean is that this is looking for `ADD(handle, offset)`, but VN's canonicalization means we can also see `ADD(offset, handle)` (it depends on prior numbering which one will it be).

Reply:
I can add it but initially I just didn't want to make code more complicated for the case that didn't raise any diffs

Reply:
> didn't raise any diffs

Is that the case? Then I think it is fine as-is. It is kind of accidental though.